lib/Controller/Schedule.php

```diff
@@ -1224,6 +1278,57 @@ public function edit($eventId)
             $this->getLog()->debug('Processed start is: FromDt=' . $fromDt->toRssString());
         }
 
+        // interrupt layout
+        if ($schedule->eventTypeId == 4) {
+
+            $campaign = $this->campaignFactory->getById($schedule->campaignId);
+
+            if ($campaign->isLayoutSpecific == 0) {
+                throw new InvalidArgumentException(__('Cannot select a campaign for this event type.'), 'campaignId');
+            }
+
+            foreach ($this->layoutFactory->getByCampaignId($schedule->campaignId) as $layout) {
+                $layoutDuration = $layout->duration;
+                $shareInSeconds = ($schedule->shareOfVoice * 3600) / 100;
+
+                if ($layoutDuration > $shareInSeconds) {
+                    throw new InvalidArgumentException(__('Layout duration is longer than share of voice in the hour period'), 'shareOfVoice');
+                }
+            }
+
+            foreach($schedule->displayGroups as $displayGroup) {
+                $events = $this->scheduleFactory->getByDisplayGroupId($displayGroup->displayGroupId);
```

Comment:
I see what you're doing here, but I don't think this is sustainable - over time we will built up more and more events on the DisplayGroup which need to be assessed. Imagine an only system with 10000 events - we go through them all here, one by one.

If we passed this events from/to date into the scheduleFactory, we could only return events here that were overlapping. It may be easier to call the `query` method directly to do that.